library/core/src/iter/adapters/dedup.rs

```diff
@@ -0,0 +1,164 @@
+    extern "rust-call" fn call_once(mut self, args: (&T, &T)) -> Self::Output {
+        (self.key)(args.0) == (self.key)(args.1)
+    }
+}
+
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+impl<F, T, K> FnMut<(&T, &T)> for ByKey<F>
+where
+    F: FnMut(&T) -> K,
+    K: PartialEq,
+{
+    #[inline]
+    extern "rust-call" fn call_mut(&mut self, args: (&T, &T)) -> Self::Output {
+        (self.key)(args.0) == (self.key)(args.1)
+    }
+}
+
+/// A zero-sized type for checking partial equality.
+///
+/// This struct acts exactly like the function [`PartialEq::eq`], but its
+/// type is always known during compile time.
+///
+/// This `struct` is created by [`Iterator::dedup`].
+/// See its documentation for more.
+///
+/// [`Iterator::dedup`]: Iterator::dedup
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+#[derive(Debug, Clone, Copy)]
+#[non_exhaustive]
+pub struct ByPartialEq;
+
+impl ByPartialEq {
+    #[inline]
+    pub(crate) fn new() -> Self {
+        Self
+    }
+}
+
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+impl<T: PartialEq> FnOnce<(&T, &T)> for ByPartialEq {
+    type Output = bool;
+    #[inline]
+    extern "rust-call" fn call_once(self, args: (&T, &T)) -> Self::Output {
+        args.0 == args.1
+    }
+}
+
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+impl<T: PartialEq> FnMut<(&T, &T)> for ByPartialEq {
+    #[inline]
+    extern "rust-call" fn call_mut(&mut self, args: (&T, &T)) -> Self::Output {
+        args.0 == args.1
+    }
+}
+
+/// An iterator that removes all but the first of consecutive elements in a
+/// given iterator satisfying a given equality relation.
+///
+/// This `struct` is created by [`Iterator::dedup`], [`Iterator::dedup_by`]
+/// and [`Iterator::dedup_by_key`]. See its documentation for more.
+///
+/// [`Iterator::dedup`]: Iterator::dedup
+/// [`Iterator::dedup_by`]: Iterator::dedup_by
+/// [`Iterator::dedup_by_key`]: Iterator::dedup_by_key
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+#[derive(Debug, Clone)]
+pub struct Dedup<I, F>
+where
+    I: Iterator,
+{
+    inner: I,
+    same_bucket: F,
+    last: Option<Option<I::Item>>,
+}
+
+impl<I, F> Dedup<I, F>
+where
+    I: Iterator,
+{
+    #[inline]
+    pub(crate) fn new(inner: I, same_bucket: F) -> Self {
+        Self { inner, same_bucket, last: None }
+    }
+}
+
+#[unstable(feature = "iter_dedup", reason = "recently added", issue = "83748")]
+impl<I, F> Iterator for Dedup<I, F>
+where
+    I: Iterator,
+    F: FnMut(&I::Item, &I::Item) -> bool,
+{
+    type Item = I::Item;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        let Self { inner, last, same_bucket } = self;
+        let last = last.get_or_insert_with(|| inner.next());
+        let last_item = last.as_ref()?;
+        let next = inner.find(|next_item| !(same_bucket)(last_item, next_item));
```

Comment:
This argument order is inconsistent with `Vec::dedup_by` (see #43041).

```suggestion
        let next = inner.find(|next_item| !(same_bucket)(next_item, last_item));
```

Reply:
You're right, thank you!